### PR TITLE
lambda.part3.exercise2 (Seliukov)

### PR DESCRIPTION
**Pull request by Duelist256**

### CODE REVIEW

File: src/test/java/lambda/part3/exercise/Exercise2.java [hunk 1]
```diff
@@ -34,51 +37,66 @@ private MapHelper(List<T> source) {
...
+                lists.add(flatMapping.apply(elem));
+            }
+
+            List<R> result = new ArrayList<>();
+            for (List<R> list : lists) {
+                result.addAll(list);
+            }
+
+            return new MapHelper<>(result);
         }
```
Comment (a-filipp0v) [inline src/test/java/lambda/part3/exercise/Exercise2.java]: Lines 59-59 could be shortened to 3 lines with forEach, andThen and method references.

File: src/test/java/lambda/part3/exercise/Exercise2.java [hunk 2]
```diff
@@ -34,51 +37,66 @@ private MapHelper(List<T> source) {
...
-        // TODO                        .flatMap(Employee -> JobHistoryEntry)
-        // TODO                        .map(JobHistoryEntry -> String(position))
-        // TODO                        .flatMap(String -> Character(letter))
-        // TODO                        .map(Character -> Integer(code letter)
-        // TODO                        .getMapped();
+        List<Integer> codes = MapHelper.from(employees)
+        .flatMap(Employee::getJobHistory)
+        .map(JobHistoryEntry::getPosition)
+        .flatMap(position -> position.chars().boxed().collect(Collectors.toList()))
+        .map(chars-> chars)
```
Comment (a-filipp0v) [inline src/test/java/lambda/part3/exercise/Exercise2.java]: that map() is unnecessary.

**Comment by Duelist256** [comment]
@a-filipp0v Have done

File: src/test/java/lambda/part3/exercise/Exercise2.java [hunk 3]
```diff
@@ -34,51 +37,61 @@ private MapHelper(List<T> source) {
          * Создает объект для маппинга, передавая ему новый список, построенный на основе исходного.
          * Для добавления в новый список каждый элемент преобразовывается с использованием заданной функции.
          * ([T], (T -> R)) -> [R]
+         *
          * @param mapping Функция преобразования элементов.
          */
         public <R> MapHelper<R> map(Function<T, R> mapping) {
-            throw new UnsupportedOperationException();
+            List<R> result = new ArrayList<>();
+            for (T elem : source) {
```
Comment (elefus) [inline src/test/java/lambda/part3/exercise/Exercise2.java]: Почему в методе `flatMap` используется forEach, а здесь `iter`?

Comment (Duelist256) [reply src/test/java/lambda/part3/exercise/Exercise2.java]: Have done

**Pull request merged**